dynamic_instrumentation/src/main.rs

```diff
@@ -0,0 +1,413 @@
+
+    pub fn from_rustc_args(at_args: &[String]) -> eyre::Result<Self> {
+        let args = rustc_driver::args::arg_expand_all(at_args);
+        let matches = rustc_driver::handle_options(&args)
+            .ok_or_else(|| eyre!("failed to parse `rustc` args"))?;
+        let session_options = rustc_session::config::build_session_options(&matches);
+        let crate_target = CrateTarget::from_session_options(session_options, &matches.free)?;
+        Ok(crate_target)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct InstrumentInfo {
+    crate_targets: Vec<CrateTarget>,
+    metadata: PathBuf,
+}
+
+fn get_sysroot_fast() -> Option<PathBuf> {
+    let sysroot = [
+        env::var_os("RUSTUP_HOME")?,
+        "toolchains".into(),
+        env::var_os("RUSTUP_TOOLCHAIN")?,
+    ]
+    .into_iter()
+    .collect();
+    Some(sysroot)
+}
+
+fn get_sysroot_slow() -> eyre::Result<PathBuf> {
+    todo!("use `rustc --print sysroot` to support non-`rustup` cases, which @fw-immunant uses")
+}
+
+fn get_sysroot() -> eyre::Result<PathBuf> {
+    get_sysroot_fast()
+        .ok_or(())
+        .or_else(|()| get_sysroot_slow())
+}
+
+fn passthrough_rustc(at_args: &[String], sysroot: &Path) -> eyre::Result<ExitStatus> {
+    // We can skip a `rustup` `rustc` -> real `rustc` resolution by just invoking the real `rustc` ourselves.
+    // TODO(kkysen) this might not work without `rustup`
+    let rustc = sysroot.join("bin/rustc");
+    let status = Command::new(rustc).args(at_args.iter().skip(1)).status()?;
+    Ok(status)
+}
+
+fn instrument_rustc(mut at_args: Vec<String>, sysroot: &Path, metadata: &Path) -> eyre::Result<()> {
+    // Normally, `rustc` looks up the sysroot by the location of its own binary.
+    // This works because the `rustc` on `$PATH` is actually `rustup`,
+    // and `rustup` invokes the real `rustc`, which is in a location relative to the sysroot.
+    // As we invoke `rustc_driver` directly here, we are `rustc`,
+    // and thus we have to explicitly specify the sysroot that the real `rustc` would normally use.
+    //
+    // Note that the sysroot contains the toolchain and host target name,
+    // but this has no effect on cross-compiling.
+    // Every toolchain's `rustc` is able to itself cross-compile.
+    // I'm not sure why the host target needs to be in the sysroot directory name, but it is.
+    //
+    // Also note that this sysroot lookup should be done at runtime,
+    // not at compile-time in the `build.rs`,
+    // as the toolchain locations could be different
+    // from where this binary was compiled and where it is running
+    // (it could be on a different machine with a different `$RUSTUP_HOME`).
+    let sysroot: &Utf8Path = sysroot.try_into()?;
+    at_args.extend(["--sysroot".into(), sysroot.as_str().into()]);
+    RunCompiler::new(&at_args, &mut MirTransformCallbacks)
+        .run()
+        .map_err(|_| eyre!("`rustc` failed"))?;
+    INSTRUMENTER.finalize(metadata).map_err(|e| eyre!(e))?;
```

Comment:
You shouldn't need to `map_err()` here, `?` should be able to convert anything that implements the `Error` trait.

Reply:
Yeah, that was due to the `anyhow` to `eyre` change (it didn't work without this conversion), but I can revert back to `anyhow` and later change to `eyre`, but change throughout the whole crate so there are no weird conversions like this.

Reply:
Fixed in https://github.com/immunant/c2rust/pull/554/commits/d8c0b4922fc5f7a38e3a81381eb83bf4fd8afab1 after reverting back to `anyhow`.